neighsyncd/neighsync.cpp

```diff
@@ -11,12 +11,244 @@
 #include "linkcache.h"
 
 #include "neighsync.h"
+#include "warm_restart.h"
 
 using namespace std;
 using namespace swss;
 
-NeighSync::NeighSync(DBConnector *db) :
-    m_neighTable(db, APP_NEIGH_TABLE_NAME)
+const NeighRestartAssit::cache_state_map NeighRestartAssit::cacheStateMap =
+{
+    {STALE,     "STALE"},
+    {SAME,      "SAME"},
+    {NEW,       "NEW"},
+    {DELETE,    "DELETE"},
+    {UNKNOWN,   "UNKNOWN"}
+};
+
+NeighRestartAssit::NeighRestartAssit(RedisPipeline *pipelineAppDB,
+  RedisPipeline *pipelineConfDB):
+    m_appTable(pipelineAppDB, APP_NEIGH_TABLE_NAME, false),
+    m_cfgTable(pipelineConfDB, CFG_WARM_RESTART_TABLE_NAME, false)
+{
+
+    m_appTableName = m_appTable.getTableName();
+
+    if (!WarmStart::isWarmStart())
+    {
+        m_warmStartInProgress = false;
+    }
+    else
+    {
+        uint32_t temp_value = WarmStart::getWarmStartTimer("neighsyncd");
+        if (temp_value != 0)
+        {
+            m_reconcileTimer = temp_value;
+        }
+    }
+}
+
+NeighRestartAssit::~NeighRestartAssit()
+{
+}
+
+/* join the field-value strings for straight printing */
+string NeighRestartAssit::joinVectorString(const vector<FieldValueTuple> &fv)
+{
+    string s;
+    for (const auto &temps : fv )
+    {
+	   s += temps.first + ":" + temps.second + ", ";
+    }
+    return s;
+}
+
+void NeighRestartAssit::setCacheEntryState(std::vector<FieldValueTuple> &fvVector,
+    cache_state_t state)
+{
+    fvVector.back().second = cacheStateMap.at(state);
+}
+
+NeighRestartAssit::cache_state_t NeighRestartAssit::getCacheEntryState(const std::vector<FieldValueTuple> &fvVector)
+{
+    for (auto &iter : cacheStateMap)
+    {
+        if (fvVector.back().second == iter.second)
+        {
+            return iter.first;
+        }
+    }
+    assert("Cache Entry State is corrupted" == NULL);
+    return UNKNOWN;
```

Comment:
>return [](start = 4, length = 6)

throw a logic_error/runtime exception?

Reply:
This is to capture the code level issue not really for run time issue like IO/resource related.

Reply:
Then 'logic_error' may fit the purpose?

---
In reply to: [214507371](https://github.com/Azure/sonic-swss/pull/599#discussion_r214507371) [](ancestors = 214507371)

Reply:
This particular error should never happen as it is just a map from string to enum. If it ever happened due to internal code error, there is nothing for the caller to do, so I think assert is the right approach here.

Reply:
No, assert is never the right approach for runtime code. In release build, assert will be optimized to nothing, and the code will silently ignore the 'logic error', and keep continue.
If you think 'there is nothing for the caller to do', you should crash or report error and exit.

Reply:
In my case, the assert is only for debug purpose for development, it is ok to remove them in release build.  But do a throw should work fine too.